Integrated Rise and DPIA type and kinding systems

michel-steuwer opened the pull request:
This PR removed the `DataType` and `Kinds` from DPIA and uses the equivalent types from Rise.

Bastacyclop reviewed (commented):
I think that it would be good to document the difference between `FragmentKind.Identifier` and `FragmentIdentifier`, both names look like synonyms to me. Maybe with a small comment on `FragmentKind.Identifier` explaining that it is a wrapper. Or maybe it could be named more appropriately (`AnyIdentifier.Fragment`?).

michel-steuwer commented on src/main/scala/rise/core/types/Solution.scala:
```diff
@@ -30,10 +31,10 @@ object Solution {
 
 case class Solution(ts: Map[Type, Type],
-                    ns: Map[NatIdentifier, Nat],
+                    ns: Map[Nat, Nat],
```
This was due to combining two substitution implementations which resulted in a change in the API.
I have reverted this API change.

michel-steuwer commented on src/main/scala/rise/autotune/package.scala:
```diff
@@ -28,7 +28,7 @@ package object autotune {
                    hierarchicalHM: Boolean = false)
 
-  case class Sample(parameters: Map[NatIdentifier, Nat],
+  case class Sample(parameters: Map[Nat, Nat],
```
Same as above, I have reverted this API change.

michel-steuwer commented on src/main/scala/rise/core/types/Type.scala:
```diff
@@ -17,147 +17,136 @@ final case class TypeIdentifier(name: String) extends Type {
-final case class OpaqueType(name: String) extends DataType {
-  override def toString: String = name
-}
+object DataType {
```
This is not necessary, but I think it is useful if we want to think about automatically generating the data type implementation (which I am) and then achieve appropriate encapsulation.

michel-steuwer commented:
@Bastacyclop Thanks for the review. I have addressed your comments.

> I think that it would be good to document the difference between `FragmentKind.Identifier` and `FragmentIdentifier`, both names look like synonyms to me. Maybe with a small comment on `FragmentKind.Identifier` explaining that it is a wrapper. Or maybe it could be named more appropriately (`AnyIdentifier.Fragment`?).

I have renamed the nested `*Kind.Identifier` classes to `*Kind.IDWrapper` to clarify their use.
I think we might be able to revisit this design when we move to Scala 3 and can use a union type to describe all possible variations of identifiers instead of these wrappers.
